timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -75,7 +96,9 @@ void BDExchangeMove<RealType>::move_device(
 
     logsumexp_.sum_device(num_target_mols_, d_log_weights_before_.data, d_log_sum_exp_before_.data, stream);
 
-    const int num_samples = 1;
+    dim3 atom_by_atom_grid(ceil_divide(N, tpb), mol_size_, 1);
+
+    const int num_samples = NUM_SAMPLES;
     for (int move = 0; move < proposals_per_move_; move++) {
```

Comment:
General question - which kernel takes up the most time for this set of kernels?

Reply:
Timings for the kernels for 2500 steps per move

The slowest is the `k_set_sampled_weight`

```
Time (%)  Total Time (ns)  Instances  Avg (ns)   Med (ns)   Min (ns)  Max (ns)  StdDev (ns)                                                  Name                                                
 --------  ---------------  ---------  ---------  ---------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     17.5       26,416,574      2,500   10,566.6   10,640.0     8,768   425,408     10,071.8  void timemachine::k_set_sampled_weight<float, (int)512>(int, int, int, const int *, const int *, co…
     13.0       19,620,340      5,000    3,924.1    4,032.0     3,359    13,280        368.8  void timemachine::k_atom_by_atom_energies<float>(int, int, const int *, const double *, const doubl…
      7.5       11,289,498      2,500    4,515.8    4,672.0     3,839    12,576        418.8  void timemachine::k_rotate_and_translate_mols<float>(int, const double *, const double *, const int…
      7.3       10,983,198      2,500    4,393.3    4,543.0     3,807    11,968        338.8  void timemachine::k_adjust_weights<float, (bool)1>(int, int, int, const int *, const int *, const T…
      7.3       10,977,489      2,500    4,391.0    4,543.0     3,807    16,864        418.8  void timemachine::k_adjust_weights<float, (bool)0>(int, int, int, const int *, const int *, const T…
      7.2       10,863,049      5,000    2,172.6    2,208.0     1,727    12,737        280.3  void gen_sequenced<curandStateXORWOW, float, int, &curand_uniform_noargs<curandStateXORWOW>, rng_co…
      4.9        7,448,702      2,500    2,979.5    3,072.0     2,591    12,896        295.0  void cub::DeviceReduceSingleTileKernel<cub::DeviceReducePolicy<cub::KeyValuePair<int, float>, cub::…
      4.7        7,129,433      2,501    2,850.6    2,880.0     2,400    12,864        663.2  void cub::DeviceReduceSingleTileKernel<cub::DeviceReducePolicy<float, float, int, cub::Max>::Policy…
      4.6        6,886,243      2,501    2,753.4    2,816.0     2,399     6,560        201.6  void cub::DeviceReduceSingleTileKernel<cub::DeviceReducePolicy<float, float, int, cub::Sum>::Policy…
      4.1        6,240,954      2,500    2,496.4    2,560.0     2,175     5,344        184.1  timemachine::k_setup_sample_atoms(int, int, const int *, const int *, const int *, int *, int *)    
      3.9        5,834,797      2,500    2,333.9    2,400.0     1,855     7,680        266.4  void gen_sequenced<curandStateXORWOW, float2, normal_args_st, &curand_normal_scaled2<curandStateXOR…
      3.5        5,202,291      2,499    2,081.7    2,144.0     1,823    12,736        258.6  void timemachine::k_store_accepted_log_probability<float>(int, const T1 *, T1 *, const T1 *, T1 *, …
      3.5        5,202,259      2,500    2,080.9    2,112.0     1,792     8,160        233.3  void timemachine::k_setup_gumbel_max_trick<float>(int, const T1 *, T1 *)                            
      3.4        5,167,353      2,500    2,066.9    2,112.0     1,791     6,848        185.8  void timemachine::k_copy_kv_key<float>(int, const cub::KeyValuePair<int, T1> *, int *)              
      3.4        5,136,022      2,500    2,054.4    2,112.0     1,791     6,336        159.8  void timemachine::k_attempt_exchange_move<float>(int, const T1 *, const T1 *, const T1 *, const dou…
      3.4        5,056,284      2,501    2,021.7    2,049.0     1,759    11,360        278.1  void timemachine::k_exp_sub_max<float>(int, const T1 *, const T1 *, T1 *)                           
      0.5          781,280          1  781,280.0  781,280.0   781,280   781,280          0.0  void timemachine::k_compute_nonbonded_target_atom_energies<float, (int)64>(int, int, const int *, c…
      0.2          269,760          2  134,880.0  134,880.0   114,240   155,520     29,189.4  void generate_seed_pseudo<rng_config<curandStateXORWOW, (curandOrdering)101>>(unsigned long long, u…
```

Reply:
Because it has to be computed in a single block due to being unable to do atomic accumulation on `__int128`